### PR TITLE
Bugfix/upload via share

### LINKED ISSUE #1878: Sometimes uploads fails or queued, then don't upload.
**Summary:** 

Our friend John Lubbock says, "I wonder if you know why when I upload photos on the Commons app, sometimes they say 'Failed' or 'queued', and don't upload?".

**Device and Android version:** 
Samsung S9+
 
 **Commons app version:** 

2.8.2

**Screen-shots:** 
image

**Would you like to work on the issue?**
^_^ no.

### PR DESCRIPTION
**Pull request by ashishkumar468**
## Sometimes uploads fails or queued, then don't upload. #1878
Fixes #1878 Sometimes uploads fails or queued, then don't upload.

## Description (required)
As the app was unable to validate the MIME type of the image shared via intent or from any source, the api to upload image used to fail because the backend was smart enough to validate the same
Fixes #{GitHub issue number and title}

{Describe the changes made and why they were made.}
Added a java library to validate the image MIME type and accordingly append the image extension. 
## Tests performed (required)

Tested on {27 & Google Pixel}, with {build variant, ProdDebug}.

## Screenshots showing what changed (optional)
 NA

### CODE REVIEW

**Comment by misaochan** [comment]
Thanks for the quick PR! A couple of things:

- We need to discuss the new library that you are adding as per https://github.com/commons-app/apps-android-commons/wiki/Code-style#new-libraries . Could you please explain why you think it is beneficial to add a library for this task, as opposed to writing the file extension checking code ourselves? Is it worth the risk of additional bugs and additional complexity, and is the library reputable enough that we can be reasonably confident that any bugs with the library itself will be solved fairly quickly in future updates?
- Could you please describe in further detail the tests you did? Did you try uploading a .png via Share and succeeded?

**Comment by ashishkumar468** [comment]
@misaochan 
Why do we need to add a new  library ? -> I actually implemented a number of ways to extract the MIME type. All of them had one issue that they relied on the data received from the intent (which as i have mentioned in the issue thread was coming wrong at-least for me in Google Pixel device and emulator ). Also, the native implementations in Android do not validate the MIME type. They instead trust the source and after trying out more than 3-4 ways (some of them include URLConnection class, URI ), all with no genuine results, I could conclude that this could not be handled natively. Finally [SimpleMagic](http://256stuff.com/sources/simplemagic/) came to rescue. Please refer https://stackoverflow.com/questions/51438/getting-a-files-mime-type-in-java (the source). The library  has 83 stars and for a project like this seems quite reputable to me. Also as we can see in the source code of a library, a number of parsers and their corresponding handlers and of-course the appropriate test cases on a number of file types is there. The code base is not that small which would take a lot of time to implement on our own. Considering a major production bug/crash (and being a great supporter of code-re usability), I went with the library, considering it the best solution as of now (and I may be wrong). I am open to suggestions for the same.

Tests-> I have manually tested by uploading both png and jpeg images confirming the extensions from the file info, both of them went to success (on Google Pixel where I could produce the issue easily as mentioned in the issue thread)

File: app/src/main/java/fr/free/nrw/commons/upload/UploadService.java [hunk 1]
```diff
@@ -183,20 +186,35 @@ public int onStartCommand(Intent intent, int flags, int startId) {
...
+        InputStream tempFileInputStream=null;
+        ContentInfo contentInfo=null;
         String notificationTag = contribution.getLocalUri().toString();
 
         try {
-            //FIXME: Google Photos bug
             File file1 = new File(contribution.getLocalUri().getPath());
             fileInputStream = new FileInputStream(file1);
+            tempFileInputStream = new FileInputStream(file1);
+            if(contentInfoUtil==null){
```
Comment (misaochan) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadService.java]: Please use consistent formatting. :) E.g. see the catch block below for our usual bracket formatting

File: app/src/main/java/fr/free/nrw/commons/upload/UploadService.java [hunk 2]
```diff
@@ -222,9 +240,17 @@ private void uploadContribution(Contribution contribution) {
 
         String filename = null;
         try {
+            //try to fetch the MIME type from contentInfo first and then use the tag to do it
+            //Note : the tag has not proven trustworthy in the past
```
Comment (misaochan) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadService.java]: Good documentation! 👍

File: app/src/main/java/fr/free/nrw/commons/upload/UploadService.java [hunk 3]
```diff
@@ -222,9 +240,17 @@ private void uploadContribution(Contribution contribution) {
...
+            String mimeType;
+            if(contentInfo==null || contentInfo.getMimeType()==null){
+                mimeType=(String)contribution.getTag("mimeType");
+            }else{
```
Comment (misaochan) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadService.java]: Formatting please

**Comment by misaochan** [comment]
@ashishkumar468 Thanks for the explanation, in that case I agree that the new library is needed as more than just a convenience. :)

I tested this on a Nexus S emulator on API 27, and indeed .png uploads via Share and the app work for me, and no issues with other filetypes either. Great job! 👍 

We just need you to make a few formatting changes, and we also need 1 more person (@neslihanturan , @maskaravivek ?) to approve the library change. Then I'll merge and release the hotfix with this.

**Comment by maskaravivek** [comment]
@ashishkumar468 The build is failing for me. I was basically trying to check the method count for this build. 

Without this build, master has `5183` methods. If the library doesn't add too many methods then IMO its safe to add this library. 

screen shot 2018-10-10 at 9 32 51 pm

**Comment by ashishkumar468** [comment]
@maskaravivek Not sure about the build failure, please try cleaning the project. And yeah regarding the method count, I have checked in both the apk's one without this library and one with the library, total method counts in both is ~17 k self and 25 k referenced hardly differing in 20-30 methods. Also, there is negligible change in apk size.

**Comment by misaochan** [comment]
Thanks Ashish!

**Pull request merged**